src/lib/http-post-parser.ts

```diff
@@ -1,357 +1,415 @@
+    // this.onBeforeSendHeadersEventDetails = onBeforeSendHeadersEventDetails;
+    this.onBeforeRequestEventDetails = onBeforeRequestEventDetails;
+    this.dataReceiver = dataReceiver;
+    /*
+    console.log(
+      "HttpPostParser",
+      // onBeforeSendHeadersEventDetails,
+      onBeforeRequestEventDetails,
     );
-    this.hasheaders = true;
-    this.body = -1;
-  } else {
-    // Let's keep an eye on unknown stream types, though we haven't seen any other stream types in our tests.
-    loggingDB.logDebug("POST request parser: unknown stream type");
-  }
-};
-
-HttpPostParser.prototype.rewind = function() {
-  this.seekablestream.seek(0, 0);
-};
-
-HttpPostParser.prototype.tell = function() {
-  return this.seekablestream.tell();
-};
-
-HttpPostParser.prototype.readLine = function() {
-  let line = "";
-  const size = this.stream.available();
-  for (let i = 0; i < size; i++) {
-    const c = this.stream.read(1);
-    if (c === "\r") {
-      continue;
-    } else if (c === "\n") {
-      break;
-    } else {
-      line += c;
-    }
+    */
   }
-  return line;
-};
-
-HttpPostParser.prototype.extractUploadStreamHeaders = function() {
-  /*
-   * Extract "Request Headers From Upload Stream" from the POST request body.
-   * Unlike standard HTTP headers, these headers are sent in the request body
-   * and contain metadata such as the size and the encoding of the submitted data
-   * (e.g. multi-part form data)
-   * Also, seee http://stackoverflow.com/questions/16548517/what-is-request-headers-from-upload-stream
-   * A sample POST request body with upload stream headers (from the `test_record_post_data_multipart_formdata`)
-   *
-    Content-Type: multipart/form-data; boundary=---------------------------1809321333852408290275809649
-    Content-Length: 455
-
-    -----------------------------1809321333852408290275809649
-    Content-Disposition: form-data; name="email"
-    [snipped]
+
+  /**
+   * @param encodingType from the HTTP Request headers
    */
-  if (this.hasheaders) {
-    this.rewind();
-    let line = this.readLine();
-    while (line) {
-      const keyValue = line.match(/^([^:]+):\s?(.*)/);
-      if (keyValue) {
-        this.postHeaders[keyValue[1]] = keyValue[2];
-      } else {
-        this.postLines.push(line);
-      }
-      line = this.readLine();
+  public parsePostRequest(encodingType) {
+    // const requestHeaders = this.onBeforeSendHeadersEventDetails.requestHeaders;
+    const requestBody = this.onBeforeRequestEventDetails.requestBody;
+    if (requestBody.error) {
+      this.dataReceiver.logError(
+        "Exception: Upstream failed to parse POST: " + requestBody.error,
+      );
+    }
+    if (requestBody.formData) {
+      return {
+        // TODO: requestBody.formData should probably be transformed into another format
+        post_body: requestBody.formData,
+      };
     }
-    this.body = this.tell();
+
+    // TODO: Refactor to corresponding webext logic or discard
```

Comment:
@englehardt: This stream-based parsing code is left - dysfunctional - in case it is relevant to use it for cases where only requestBody.raw is supplied. Please remove if not relevant.

Reply:
I suspect a lot of the code can be removed. It's great that WebExtensions handles all the stream parsing! The right thing to do will be to check how the `requestBody.raw` property looks for the various OpenWPM test pages (https://github.com/citp/OpenWPM/blob/master/test/test_http_instrumentation.py#L527).